### ISSUE TITLE
Add support for handlers to react to internal state changes.

### ISSUE THREAD

**Issue by nova4u** [measuredco/puck#534]
### Description

The existing `onChange` callback in Puck is too broad and lacks specificity. It's triggered in a useEffect hook whenever the data inside Puck changes, without providing any context about the type of event that occurred or what specifically changed. This approach has several drawbacks:

- It requires manual searching and comparison to determine what actually changed.
- It triggers after the data has already been modified, limiting real-time responsiveness.
- It doesn't distinguish between different types of actions (e.g., insertions, deletions, moves).
- It provides no information about the components involved in the change.

Propose to add onInsert, onRemove and other events. These handlers will allow users to respond to component insertion and removal events.

I already have a working implementation and would be happy to open a PR.

### Rationale

Currently, Puck doesn't provide a way for users to hook into the component insertion and removal processes. Adding these handlers will:

- Enhance the flexibility of the editor by allowing custom logic to be executed when components are added or removed.
- Enable better integration with external systems or state management solutions.
- Provide opportunities for analytics, logging, or other side effects tied to editing actions.

### Proposed Changes

1. Update the createReducer function to accept handlers.
2. Modify the storeInterceptor to call these handlers.
3. Update the Puck component to accept onInsert and onRemove as optional props.

```typescript
<Puck
  onInsert={(newItem, zone, index) => {
    ...
  }}
  onRemove={(removedItem, zone, index) => {
  ...
  }}
  ...
/>
`

**Comment by chrisvxd** [comment]
Thanks @nova4u, and great proposal. I do wonder if it might be better to expose the actions to the user and expose the entire state machine, rather than adding an `onX` callback for every type of action the user might want to react to.

```tsx
<Puck
  onAction={(action, appState, prevAppState) => {
    if (action.type === "insert") {
      // do something
    }
  }}
/>
```

**Comment by chrisvxd** [comment]
Closed by #538, will be available in next canary.
